### PR TITLE
[AArch64] Implement fcmp{u,o}

### PR DESCRIPTION
**Pull request by Sonicadvance1**

### CODE REVIEW

File: Source/Core/Core/PowerPC/JitArm64/JitArm64_FloatingPoint.cpp [hunk 1]
```diff
@@ -392,3 +392,90 @@ void JitArm64::fsubx(UGeckoInstruction inst)
...
+	ORR(XA, XA, 64 - 63, 0, true);
+	continue1 = B();
+
+	if (a != b)
+	{
+		SetJumpTarget(pNaN1);
+		SetJumpTarget(pNaN2);
+	}
+	else
+		SetJumpTarget(pNaN3);
```
Comment (lioncash) [inline Source/Core/Core/PowerPC/JitArm64/JitArm64_FloatingPoint.cpp]: missing braces

File: Source/Core/Core/PowerPC/JitArm64/JitArm64_FloatingPoint.cpp [hunk 2]
```diff
@@ -392,3 +392,92 @@ void JitArm64::fsubx(UGeckoInstruction inst)
...
+	if (a == b)
+		pNaN3 = B(CC_NEQ);
+
+	if (a != b)
+	{
+		// if B > A goto Greater's jump target
+		pGreater = B(CC_GT);
+	}
+
+	if (a != b)
```
Comment (degasus) [inline Source/Core/Core/PowerPC/JitArm64/JitArm64_FloatingPoint.cpp]: No need for this second if clause, just merge it with the last one.

File: Source/Core/Core/PowerPC/JitArm64/JitArm64_FloatingPoint.cpp [hunk 3]
```diff
@@ -392,3 +392,92 @@ void JitArm64::fsubx(UGeckoInstruction inst)
...
+	{
+		// if B < A, goto Lesser's jump target
+		pLesser = B(CC_MI);
+	}
+
+	ORR(XA, XA, 64 - 63, 0, true);
+	continue1 = B();
+
+	if (a != b)
```
Comment (degasus) [inline Source/Core/Core/PowerPC/JitArm64/JitArm64_FloatingPoint.cpp]: This NaN part could be in the farcode, so we'll save the continue1 branch if a==b

File: Source/Core/Core/PowerPC/JitArm64/JitArm64_FloatingPoint.cpp [hunk 4]
```diff
@@ -392,3 +392,92 @@ void JitArm64::fsubx(UGeckoInstruction inst)
...
+
+	m_float_emit.FCMP(EncodeRegToDouble(VA), EncodeRegToDouble(VB));
+
+	if (a == b)
+		pNaN3 = B(CC_NEQ);
+
+	if (a != b)
+	{
+		// if B > A goto Greater's jump target
+		pGreater = B(CC_GT);
```
Comment (degasus) [inline Source/Core/Core/PowerPC/JitArm64/JitArm64_FloatingPoint.cpp]: Not quite sure if this is better:
```
pnGreater = B(CC_LE);
ORR(XA, XA, 0, 0, true);
continueX = B();
SetJumpTarget(pnGreater);
```

File: Source/Core/Core/PowerPC/JitArm64/JitArm64_FloatingPoint.cpp [hunk 5]
```diff
@@ -392,3 +392,92 @@ void JitArm64::fsubx(UGeckoInstruction inst)
...
+	{
+		SetJumpTarget(pNaN1);
+		SetJumpTarget(pNaN2);
+	}
+	else
+	{
+		SetJumpTarget(pNaN3);
+	}
+
+	ORR(XA, XA, 64 - 61, 0, true);
```
Comment (degasus) [inline Source/Core/Core/PowerPC/JitArm64/JitArm64_FloatingPoint.cpp]: tbh, I doubt this NaN code is near to complete, but good luck, it only happens rarely. What do you think about a conditional fallback to interpreter in far code?

**Comment by Sonicadvance1** [comment]
I think it's good enough as is at the moment.
We can't do a conditional fixupbranch farther away than 1MB so it's difficult to move bits of it to farcode without adding a lot more branches, and these little Cortex-A57s take performance penalties if there are more than 1 branch instruction within 16 bytes of each other. So the less branches the better here.

If someone wants to improve it in the future that would be great, I'm busy working on other things atm.

**Comment by Sonicadvance1** [comment]
Changing this to depend on PR #2837, so that will need to be merged before this one.

File: Source/Core/Core/PowerPC/JitArm64/JitArm64_FloatingPoint.cpp (hunk 3 above)
Comment (Sonicadvance1) [inline Source/Core/Core/PowerPC/JitArm64/JitArm64_FloatingPoint.cpp]: Problem is that conditional branches can only branch +-1MB of the current PC, and the farcode is quite a bit farther away than that. Would need some re-imagining.

File: Source/Core/Core/PowerPC/JitArm64/JitArm64_FloatingPoint.cpp (hunk 4 above)
Comment (Sonicadvance1) [inline Source/Core/Core/PowerPC/JitArm64/JitArm64_FloatingPoint.cpp]: Don't think it really matters in this instance.
All these branches are going to be fairly terrible no matter what.

**Comment by Sonicadvance1** [comment]
I'll most likely come back to optimize this in the future. I still think there are some things that can be done, but currently it is better purely to have it implemented than not.

**Comment by Sonicadvance1** [comment]
This actually doesn't depend on the other PR due to how that PR is implemented.

**Pull request merged**